### PR TITLE
new:(lib) distgen lib to gen distributable bins

### PR DESCRIPTION
**Pull request by afiune**
## Description
A simple generator for creating easily distributable Go packages.

### Simple usage:
```go
package main
//go:generate go run github.com/chef/go-libs/distgen
```
Look at [distgen/README.md](distgen/README.md) for more examples.

Signed-off-by: Salim Afiune <afiune@chef.io>

## Related Issue
Source: https://github.com/afiune/godist
Upstream Dep: https://github.com/chef/chef-analyze/pull/67

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Chore (non-breaking change that does not add functionality or fix an issue)

## Checklist:

- [x] I have read the **CONTRIBUTING** document.
- [x] I have run the pre-merge tests locally and they pass.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] All commits have been signed-off for [the Developer Certificate of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -60,3 +60,13 @@ func main() {
...
+
+## `distgen`
+A simple generator for creating easily distributable Go packages.
+
+### Simple usage:
+```go
+package main
+//go:generate go run github.com/chef/go-libs/distgen
+```
+Look at [distgen/README.md](distgen/README.md) for more example.
```
Comment (mjingle) [inline README.md]:
```suggestion
Look at the [distgen README](distgen/README.md) for more examples.
```

File: distgen/README.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
+# distgen
+A simple generator for creating easily distributable Go packages.
+
+This automation provides an easy way to generate variables that can be configured
+globally across multiple go packages, such as trademarks, product names, websites,
+etc. This generator should be defined as a `go:generate` comment and run at build
+time, using the `go generate` command.
+
+### Example-1
```
Comment (mjingle) [inline distgen/README.md]:
```suggestion
### Single Main Package Example
```

File: distgen/README.md [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+```go
+package main
+//go:generate go run github.com/chef/go-libs/distgen
+```
+
+The automation will deploy a file called `dist.go` with all the variables defined
+inside the JSON file `glob_dist.json` inside this repository. (See a real example
+at [example-main/](example-main).)
+
+### Example-2
```
Comment (mjingle) [inline distgen/README.md]:
```suggestion
### Multipackage Example
```

File: distgen/README.md [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+```
+
+This usage is for go projects that has multiple packages. By creating a single `dist`
+package inside your repository, you can import the generated package in any other
+packages. (See a real example at [example-multi-pkg/](example-multi-pkg).)
+
+### Example-3
+
+To fully customize this automation, a user can provide a URL pointing to a custom JSON
+file as a third parameter to the `go:generate` directive. This custom JSON file should contain the
```
Comment (mjingle) [inline distgen/README.md]:
```suggestion
file as a third parameter of the `go:generate` directive. This custom JSON file should contain the
```

File: distgen/README.md [hunk 5]
```diff
@@ -0,0 +1,45 @@
+# distgen
+A simple generator for creating easily distributable Go packages.
+
+This automation provides an easy way to generate variables that can be configured
+globally across multiple go packages, such as trademarks, product names, websites,
+etc. This generator should be defined as a `go:generate` comment and run at build
+time, using the `go generate` command.
+
+### Example-1
+Our first example involves a simple use within a single main package. First, create
```
Comment (mjingle) [inline distgen/README.md]:
```suggestion
Our first example involves a simple use within a single `main` package. First, create
```

File: distgen/README.md [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+```go
+package dist
+//go:generate go run github.com/chef/go-libs/distgen global.go dist
+```
+
+This usage is for go projects that has multiple packages. By creating a single `dist`
+package inside your repository, you can import the generated package in any other
+packages. (See a real example at [example-multi-pkg/](example-multi-pkg).)
+
+### Example-3
```
Comment (mjingle) [inline distgen/README.md]: Would this be the correct heading based on the content below?
```suggestion
### Custom JSON File Example
```

File: distgen/main.go [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+	// a JSON format file where the global variables are defined
+	// (by default the automation use the local `glob_dist.json`)
+	//
+	// *customizable* (args:3)
+	globDistJson = "https://raw.github.com/afiune/godist/master/glob_dist.json"
+	// chicken and egg problem! when we update this file, we need to first merge the
+	// change and then, run the go generation as a following step, maybe we can use
+	// expeditor here but for now, I will merge it with my json and then point to
+	// the real file below:
+	//globDistJson = "https://raw.github.com/chef/go-libs/master/distgen/glob_dist.json"
```
Comment (afiune) [inline distgen/main.go]: I'll open a new PR to update this code! 👍

**Pull request merged**